File: algorithms/206.c
#include <stdio.h>

struct ListNode
{
    int val;
    struct ListNode *next;
};

struct ListNode* reverseList(struct ListNode* head)
{
    if ( !head )
        return head;
    struct ListNode *p, *q;
    p = head->next;
    head->next = NULL;
    while ( p )
    {
        q = p;
        p = p->next;
        q->next = head;
        head = q;
    }
    return head;
}

int main()
{
    return 0;
}

File: algorithms/python/206.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head: return None

        def helper(head):
            if not head.next:
                return head, head
            node, head = head, head.next
            node.next = None
            head, tail = helper(head)
            tail.next = node
            return head, node
        head, tail = helper(head)
        return head
